### DISCUSSION TITLE
Documents unclear on fetching multiple images/files from data array.

### DISCUSSION THREAD

**Discussion by ghost** [pocketbase/pocketbase#473] [Q&A] [answered]
In the official JavaScript SDK documentation there is a **File URL** section that lists and gives an example on how to fetch a file from the database itself. However this example is not very clear for some use cases :s 

Getting a single image from a single record is depicted on those examples, however there is no clear way on how to get the image url's for multiple records:

Let's suppose we want to fetch all of our cars which have a name and a photo in svelte:

```svelte
const records = = await client.records.getFullList('cars', 200, {
            sort: '-created',
        });
```

If we do a `console.log(records)` we get an output like this:

```json
[
  {
    "id": "owtmu5ub03v2jnv",
    "created": "2022-09-15 03:48:45.807",
    "updated": "2022-09-15 03:48:45.807",
    "@collectionId": "dcw23mvm7dfan0i",
    "@collectionName": "cars",
    "imagen_destacada": "car1.png",
    "nombre": "Nissan",
    "@expand": {}
  },
  {
    "id": "da8g3eqfgjptlds",
    "created": "2022-09-15 03:48:25.170",
    "updated": "2022-09-15 03:48:25.170",
    "@collectionId": "dcw23mvm7dfan0i",
    "@collectionName": "cars",
    "imagen_destacada": "car2.png",
    "nombre": "Mercedes",
    "@expand": {}
  },
  {
    "id": "yif6vn5akjobo0p",
    "created": "2022-09-15 03:47:58.068",
    "updated": "2022-09-15 03:47:58.068",
    "@collectionId": "dcw23mvm7dfan0i",
    "@collectionName": "cars",
    "imagen_destacada": "car3.png",
    "nombre": "Ford",
    "@expand": {}
  },
  {
    "id": "zxmxcnp4r3d94p6",
    "created": "2022-09-15 03:36:42.169",
    "updated": "2022-09-15 03:47:31.644",
    "@collectionId": "dcw23mvm7dfan0i",
    "@collectionName": "cars",
    "imagen_destacada": "car4.png",
    "nombre": "Mini",
    "@expand": {}
  }
]
```

If you go to your svelte component and try to "print and style" those cars you'll face a funny error:
Let's suppose this is our component

```html
{#each record as car}
<div class="container">
    <img src={car.imagen_destacada}/>
    <p>{car.nombre}</p>
</div>
{/each}
```

You'll get a 404 error when trying to access the image, namely: `GEThttp://localhost:5173/498px_svelte_logo_svg_e1648074191362_I0q6vc9ckv.png
[HTTP/1.1 404 Not Found 2ms]`

Problem is how to access these image URL's in the array so svelte can fetch them properly? The snippet included in the docs:

```js
const firstFilename = record.documents[0];

// http://127.0.0.1:8090/api/files/demo/kfzjt5oy8r34hvn/test_52iWbGinWd.png?thumb=100x250
const url = client.records.getFileUrl(record, firstFilename, {'thumb': '100x250'});
```

doesn't fullfill it's purpose correctly. As we don't want just one image of one element, we eant all the image URLS.

I've tried manually building the API URL, in the templates to avoid confusion. Let's suppose you defined a variable for your api url named `api_url` that points to `api.domain.com`.

```html
{#each record as car}
<div class="container">
    <img src={api_url + car.imagen_destacada}/>
    <p>{car.nombre}</p>
</div>
{/each}
```

This wont work as well as the full valid URL to get a record is `http://api.domain.com/api/files/COLLECTION_ID_OR_NAME/RECORD_ID/FILENAME` so, here comes the second part.

Okay, the collection ID and name are in the JSON response so we can get them...oh wait...what about their notation?
```json
    "@collectionId": "dcw23mvm7dfan0i",
    "@collectionName": "cars",
```

That's a huge problem since Svelte text expressions are kind of...well..garbage and very limited so accessing those in any form will result in an "Unexpected Token Error":

Without quotes:
```html
{#each record as car}
<div class="container">
    <img src={api_url + car.@collectionName + car.imagen_destacada}/>
    <p>{car.nombre}</p>
</div>
{/each}
```

With quotes:
```html
{#each record as car}
<div class="container">
    <img src={api_url + car."@collectionName" + car.imagen_destacada}/>
    <p>{car.nombre}</p>
</div>
{/each}
```
Both give an unexpected token error. 

## How to solve this?

Is there a way to fetch all the records with their respective URL's? Or is Pocketbase SDK limited to those examples? Any help would be much appreciated.
 
So far the only progress made is by setting the collection id on a string variable and using that on the templates.

**Comment by ganigeorgiev** [comment, marked as answer]
> That's a huge problem since Svelte text expressions are kind of...well..garbage and very limited so accessing those in any form will result in an "Unexpected Token Error":

This doesn't matter. You can access the object via the array notation, eg. `car["@collecitonName"]` (just for info `collecitonId` and `collectionName` will be base root level keys after the changes in #376).

---

The above cited doc example, `client.records.getFileUrl(record, firstFilename, {'thumb': '100x250'});` should work just fine:

```js
{#each record as car}
    <div class="container">
        <img src={client.records.getFileUrl(car, car.imagen_destacada)} />
        <p>{car.nombre}</p>
    </div>
{/each}
```

In any case, the documentation will be refactored as part of #376.

**Reply by ghost** [reply]
I'll try that, so far I've managed to achieve this by making a HUGE concatenation:

 ```html
<img height="50" width="60" class="img-fluid" src="{ fapi_url + "api/files/" + record['@collectionId'] + "/" + record.id + "/" + record.imagen_destacada }" alt="{record.nombre}"/>
```

I'll try that expression. In any case I'd like to contribute to the documentation for the Issues I've faced in case someone else bumps into them :face_in_clouds:  Would it be wise to wait until #376 or where can docs be submitted :) ?

**Reply by ganigeorgiev** [reply]
`client.records.getFileUrl` will do the concatenation for you.

It is better to wait #376 for doc changes to avoid applying them twice.

**Reply by ghost** [reply]
In the mean time. I'll mark the solutions in the discussions. Might as well write a blog post for spanish users. Thank you so much for your time and patience! :bowing_man:

**Reply by TraFost** [reply]
thanks bro u saved me a lot
